crates/wasmparser/src/validator/component.rs

```diff
@@ -711,19 +711,31 @@ impl ComponentState {
                 .chain(ty.results.iter().map(|(_, ty)| ty))
                 .all(|ty| types.type_named_valtype(ty, set)),
 
-            // Instances must recursively have all referenced types named.
-            Type::ComponentInstance(ty) => ty.exports.values().all(|ty| {
-                let id = match ty {
-                    ComponentEntityType::Module(id)
-                    | ComponentEntityType::Func(id)
-                    | ComponentEntityType::Value(ComponentValType::Type(id))
-                    | ComponentEntityType::Type { created: id, .. }
-                    | ComponentEntityType::Instance(id)
-                    | ComponentEntityType::Component(id) => *id,
-                    ComponentEntityType::Value(ComponentValType::Primitive(_)) => return true,
+            // Instances must recursively have all referenced types named. Resources defined by
+            // this interface are considered defined
+            Type::ComponentInstance(ty) => {
+                let mut with_resources = HashSet::new();
+
+                let set = if ty.explicit_resources.is_empty() {
```

Comment:
While a bit subtle, I think that this check may want to look at the `defined_resources` list rather than look at the `explicit_resources` map. The reason for that is that concrete instances will have an empty list there but the instance types, which this PR is fixing, may not have an empty `defined_resources` list. On the other hand though it's frequent for any kind of instance to have an `explicit_resources` array set, so this I don't think is the right branch in the implementation.

The reason for this is that the set of exported resources should be only judiciously added to and for normal instances everywhere the previous additions to the set should already be taken care of. In effect by adding all `explicit_resources` it's forcing this check to basically pass.

So, in other words, I think this check should be refactored to look more like:

```rust
                let set = if ty.defined_resources.is_empty() {
                    set
                } else {
                    with_resources = set.clone();
                    for resource in ty.defined_resources.iter() {
                        with_resources.insert(ty.type_id_of(resource));
                    }
                    &with_resources
                };
```

or something along those lines where `type_id_of` is the body of the loop in the `explicit_resources` method you added. Either that or rename the added `explicit_resources` to `defined_resources` and have the same loop you have today returning an iterator (actually after typing this out I prefer the latter style here which you already have)

Reply:
Wouldn't only looking at `defined_resources` mean that the test you mentioned above wouldn't work?
```
(component
  (type (export "x") (instance
    (export $i "i" (instance
      (export "x" (type (sub resource)))
    ))

    (alias export $i "x" (type $x))
    (export "f" (func (param "x" (borrow $x))))
  ))
)
```